File: 0563-binary-tree-tilt/0563-binary-tree-tilt.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function findTilt(root: TreeNode | null): number {
  const traverse = (node: TreeNode | null) => {
    if (node === null) return [0, 0];
    const [leftTiltSum, leftValSum] = traverse(node.left);
    const [rightTiltSum, rightValSum] = traverse(node.right);
    return [
      Math.abs(leftValSum - rightValSum) + leftTiltSum + rightTiltSum,
      leftValSum + node.val + rightValSum,
    ];
  }
  return traverse(root)[0];
};

File: 0563-binary-tree-tilt/0563-binary-tree-tilt.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findTilt(root *TreeNode) int {
  var traverse func(node *TreeNode) (int, int)
  traverse = func(node *TreeNode) (int, int) {
    if node == nil {
      return 0, 0
    }
    changedSum, nodeLeftValSum, nodeRightValSum := 0, 0, 0
    if node.Left != nil {
      var nodeLeftChangedSum int
      nodeLeftChangedSum, nodeLeftValSum = traverse(node.Left)
      changedSum += nodeLeftChangedSum
    }
    if node.Right != nil {
      var nodeRightChangedSum int
      nodeRightChangedSum, nodeRightValSum = traverse(node.Right)
      changedSum += nodeRightChangedSum
    }
    changedSum += int(math.Abs(float64(nodeLeftValSum) - float64(nodeRightValSum)))
    return changedSum, nodeLeftValSum + node.Val + nodeRightValSum
  }
  ans, _ := traverse(root)
  return ans
}